mongo_orchestration/launch.py

```diff
@@ -0,0 +1,354 @@
+            ])
+        return {'members': members}
+
+    def _init_from_response(self, response):
+        self.id = response['id']
+        self.uri = response.get('mongodb_auth_uri', response['mongodb_uri'])
+        for member in response['members']:
+            m = Server(member['server_id'], member['host'])
+            self.members.append(m)
+            if member['state'] == 1:
+                self.primary = m
+            elif member['state'] == 2:
+                self.secondary = m
+        return self
+
+    def start(self):
+        # We never need to restart a replica set, only start new ones.
+        return self._init_from_response(self._make_post_request())
+
+    def restart_primary(self):
+        self.primary.stop(destroy=False)
+        time.sleep(5)
+        self.primary.start()
+        time.sleep(1)
+        self._init_from_response(self._make_get_request())
+        print('New primary: %s' % self.primary.uri)
+
+
+class ReplicaSetSingle(ReplicaSet):
+
+    def get_config(self):
+        return {
+            'members': [
+                {'procParams': self.proc_params()}
+            ]
+        }
+
+
+class ShardedCluster(MCTestObject):
+
+    _resource = 'sharded_clusters'
+    _shard_type = ReplicaSet
+
+    def __init__(self, **kwargs):
+        self.id = None
+        self.uri = None
+        self.shards = []
+        self._proc_params = kwargs
+
+    def get_config(self):
+        return {
+            # 'configsvrs': [{'members': [DEFAULT_OPTIONS.copy()]}],
+            'routers': [self.proc_params(), self.proc_params()],
+            'shards': [
+                {'id': 'demo-set-0', 'shardParams':
+                    self._shard_type().get_config()},
+                # {'id': 'demo-set-1', 'shardParams':
+                #     self._shard_type().get_config()}
+            ]
+        }
+
+    def start(self):
+        # We never need to restart a sharded cluster, only start new ones.
+        response = self._make_post_request()
+        for shard in response['shards']:
+            shard_resp = requests.get(_mo_url('replica_sets', shard['_id']))
+            shard_json = shard_resp.json()
+            self.shards.append(self._shard_type()._init_from_response(shard_json))
+        self.id = response['id']
+        self.uri = response.get('mongodb_auth_uri', response['mongodb_uri'])
+        return self
+
+
+class ShardedClusterSingle(ShardedCluster):
+    _shard_type = ReplicaSetSingle
+
+
+def argv_has(string):
+    return any(string in arg for arg in sys.argv[1:])
+
+
+DEFAULT_CERTS = os.path.join(
+    os.environ.get(
+        'MONGO_ORCHESTRATION_HOME', os.path.dirname(__file__)),
+    'lib'
+)
+CERTS = os.environ.get('MONGO_ORCHESTRATION_CERTS', DEFAULT_CERTS)
+
+
+def main():
+    for arg in sys.argv[1:]:
+        try:
+            port = int(arg)
+            _free_port = itertools.count(port)
+        except:
+            pass
+    for version in ['2.6.12', '3.0.12', '3.2.10', '3.4.0-rc0']:
```

Comment:
-> `version in ['3.6', '4.0', '4.2', '4.4', '5.0', '6.0', '7.0', 'latest']:`

Reply:
Done